src/string_functions.rs

```diff
@@ -0,0 +1,2048 @@
+/// argument.
+/// Encoding: UTF-8
+///
+/// # Safety
+///
+/// Works on raw pointers, inherently unsafe.
+/// Will panic if trying to index outside of the array or trying
+/// to replace more characters than remaining.
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn CONCAT_EXT__STRING(
+    dest: *mut u8,
+    argc: i32,
+    argv: *const *const u8,
+) -> i32 {
+    if argv.is_null() || dest.is_null() {
+        panic!("Received null-pointer.")
+    }
+    let mut dest = dest;
+    let mut argv = argv;
+    for _ in 0..argc {
+        EncodedCharsIter::decode(*argv).encode(&mut dest);
+        argv = argv.add(1);
+    }
+
+    0
+}
+
+/// Concatenates all given strings in the order in which they are given.
+/// Strings are passed as pointer of pointer to u8, where each pointer represents
+/// the starting address of each string. The amount of strings must be passed as
+/// argument.
+/// Encoding: UTF-16
+///
+/// # Safety
+///
+/// Works on raw pointers, inherently unsafe.
+/// Will panic if trying to index outside of the array or trying
+/// to replace more characters than remaining.
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn CONCAT__WSTRING(
+    argc: i32,
+    argv: *const *const u16,
+) -> Wrapper<[u16; 2048]> {
+    if argv.is_null() {
+        panic!("Received null-pointer.")
+    }
+    let mut dest = [0_u16; 2048];
+    let mut dest_ptr = dest.as_mut_ptr();
+    let mut argv = argv;
+    // what should happen if passed parameters exceed 2048 bytes? how to check without
+    // introducing too much overhead?
+    for _ in 0..argc {
+        EncodedCharsIter::decode(*argv).encode(&mut dest_ptr);
+        argv = argv.add(1);
+    }
+
+    Wrapper { inner: dest }
+}
+
+/// Concatenates all given strings in the order in which they are given.
+/// Strings are passed as pointer of pointer to u8, where each pointer represents
+/// the starting address of each string. The amount of strings must be passed as
+/// argument.
+/// Encoding: UTF-16
+///
+/// # Safety
+///
+/// Works on raw pointers, inherently unsafe.
+/// Will panic if trying to index outside of the array or trying
+/// to replace more characters than remaining.
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn CONCAT_EXT__WSTRING(
+    dest: *mut u16,
+    argc: i32,
+    argv: *const *const u16,
+) -> i32 {
+    if argv.is_null() || dest.is_null() {
+        panic!("Received null-pointer.")
+    }
+    let mut dest = dest;
+    let mut argv = argv;
+    for _ in 0..argc {
+        EncodedCharsIter::decode(*argv).encode(&mut dest);
+        argv = argv.add(1);
+    }
+
+    0
+}
+
+/// Helper function for generic, variadic string equality functions.
+fn compare<T>(argc: i32, argv: *const *const T, predicate_func: fn(Ordering) -> bool) -> bool
+where
+    T: Ord + PrimInt,
+{
+    if argc < 2 {
+        // what is the desired behaviour for this? panic? true or false?
+        return false;
```

Comment:
I would like to say compile error, but since we can't it's either a panic or true (not false, because a single string in a collection matches all the strings in that collection)

Reply:
I would prefer the panic though